js/source/worker.js

```diff
@@ -153,7 +171,25 @@ util.extend(Worker.prototype, {
     'query features': function(params, callback) {
         var tile = this.loaded[params.source] && this.loaded[params.source][params.uid];
         if (tile) {
-            tile.featureTree.query(params, callback);
+
+            var id;
+
+            var classString = Object.keys(params.classes).join(' ');
+            if (this._cascadedClasses !== classString) {
+                this._cascadedClasses = classString;
+                for (id in this.styleLayersByID) {
+                    this.styleLayersByID[id].cascade(params.classes, {transition: false}, {});
+                }
+            }
+
+            if (this._recalculatedZoom !== params.zoom) {
+                this._recalculatedZoom = params.zoom;
+                for (id in this.styleLayersByID) {
+                    this.styleLayersByID[id].recalculate(params.zoom, { lastIntegerZoom: Infinity, lastIntegerZoomTime: 0, lastZoom: 0 });
+                }
+            }
+
+            callback(null, tile.featureTree.query(params, this.styleLayersByID));
         } else {
```

Comment:
We currently create an instance of `Style` per bucket. That now strikes me as silly. We should create a single persistent `Style` per `Worker` and update it whenever we receive `set layers`